sprint_tasks/MSI_2/MSI_send_message_ms_teams_dag_complex_dependencies.py

```diff
@@ -0,0 +1,188 @@
+        """Method combines the whole message and send it to MSTeams channel."""
+        myTeamsMessage = pymsteams.connectorcard(self.webhook_url)
+        myTeamsMessage.color(self.MESSAGE_COLOR)
+        myTeamsMessage.title(title)
+        myTeamsMessage.text(
+            f"**Quote:** {quote_text}\n\n"
+            f"**Author:** {quote_author}\n\n"
+            f"![Image]({image_url})"
+        )
+
+        try:
+            myTeamsMessage.send()
+            logging.info("Message successfully sent.")
+            return
+        except (ConnectionError, HTTPError, TimeoutError) as e:
+            logging.error(
+                f"Failed to send message due to network error: {e}",
+                exc_info=True
+            )
+        except pymsteams.TeamsWebhookException as e:
+            logging.error(
+                f"Failed to send message due to Teams webhook error: {e}",
+                exc_info=True
+            )
+        except ValueError as e:
+            logging.error(
+                f"Failed to send message due to value error: {e}",
+                exc_info=True
+            )
+
+
+def load_quote(ti):
+    quote_generator = QuoteGeneratorFromQuotableAPI()
+    quote_content = quote_generator.get_content()
+    if not quote_content:
+        logging.error("Failed to generate quote")
+        return None, None
+    quote_text, quote_author = quote_content
+    ti.xcom_push(
+        key='quote_text', 
+        value=quote_text
+    )
+    ti.xcom_push(
+        key='quote_author', 
+        value=quote_author
+    )
+
+
+def load_image(ti):
+    image_generator = ImageGeneratorFromPiscumAPI()
+    image_url = image_generator.get_content()
+    if not image_url:
+        logging.error("Failed to generate image.")
+    ti.xcom_push(
+        key='image_url', 
+        value=image_url
+    )
+
+
+def send_message(ti):
+    quote_text = ti.xcom_pull(
+        key='quote_text', 
+        task_ids='load_quote'
+    )
+    quote_author = ti.xcom_pull(
+        key='quote_author', 
+        task_ids='load_quote'
+    )
+    image_url = ti.xcom_pull(
+        key='image_url', 
+        task_ids='load_image'
+    )
+    if not quote_text or not quote_author or not image_url:
+        logging.error("Failed to send message due to missing content")
+        return
+
+    message_sender = TeamsMessageSender(TEAMS_WEBHOOK_URL)
+    message_sender.send_message(
+        "Sent by Natalia Ananeva",
+        quote_text,
+        quote_author,
+        image_url
+    )
+
+
+with DAG(
+    dag_id='send_massage_msteams_dag_complex_dependencies',
+    start_date=datetime.datetime(
+        year=2024, 
+        month=6, 
+        day=10
+    ),
+    schedule_interval="0 12 * * *",  # send every day at 12 a.m.
+    tags=['send_message_teams'],
+    description='A DAG to send message to MS Teams channel using webhook. \
+        Message consists of quote and picture. ',
+    catchup=False
+) as dag:
```

Comment:
It's actually better to explicitly fail for some tasks if something goes wrong - this way we would understand where do we have a problem, and not research logs for the whole DAG

Reply:
To explicitly fail Airflow tasks ,  it's better to catch these exceptions within  Python callables? Am I understood correctly?